### PR TITLE
feat!: Replace lilliput with govips

### LINKED ISSUE #259: Explore if we can serve image thumbnail for WebP images with reasonable effort
Example URL for which I would like a thumbnail served: https://artflowbucket-new.s3.amazonaws.com/generated/92fc3c00e617436e930cd9307974a72f.webp

This might need client-side support from Chatterino2 unless we want to convert the images before sending them

### PR DESCRIPTION
**Pull request by KararTY**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description

<details>
  <summary>Old text</summary>

Preview: https://streamable.com/0lu2pe

There's a few pain-points with govips (If your distro has libvips below version 8.12 you will have to build from source) but it solves several issues that lilliput has:

* Advertises low memory usage, https://github.com/libvips/libvips/wiki/Speed-and-memory-use
* Animated WebP is supported. (See video)
* The output parameters actually work, which means one can reduce the output size considerably. (See video)

We can do custom outputs based on the input, which also means custom quality etc. (In video everything is exported as WebP with quality at 10%)

But, I'm just danking around. Consider this PR a proof-of-concept, maybe someone more can look into govips / vips and see how we could implement this.

For building libvips, I used the following: https://github.com/libvips/libvips/wiki/Build-for-Ubuntu

</details>

This PR brings support for animated image previews without lilliput by replacing it with a library (govips) that interfaces with libvips.

Why is this PR replacing lilliput? lilliput has several issues related to memory usage, which libvips resolves as well as issues with animated WebP that are yet to be supported even on lilliput's main contributor and user: Discord. libvips resolves both issues: See contents of "Old text" for more context.

Supported animated file formats with this PR:
* GIF
* WebP ( Fixes #259 )

Not (yet) supported animated file formats:
* Animated "image/avif" [libvips/libvips/discussions/3048](https://github.com/libvips/libvips/discussions/3048) - Couldn't get it to work on Windows or WSL/Ubuntu. See comment: https://github.com/Chatterino/api/pull/312#issuecomment-1281586330

All inputs are currently turned into WebP to potentially save on bandwidth. One can save more bandwidth by setting a quality limit on the WebP output, **but for this PR I decided to leave this out: It can be its own PR with a new config entry.**

Additionally, this PR introduces another breaking change by renaming `enable-lilliput` config entry to `enable-animated-thumbnails`.

<details>
  <summary>Test links</summary>

Animated WebP https://7tv.app/emotes/604281c81ae70f000d47ffd9
Animated WebP https://mathiasbynens.be/demo/animated-webp-supported.webp
Static WebP https://7tv.app/emotes/617142c4ffc7244d797caa57
Static WebP https://www.gstatic.com/webp/gallery/5.webp 
Animated GIF https://codelabs.developers.google.com/static/codelabs/avif/images/ice_qcif_15fps.gif
Animated GIF https://media.tenor.com/2hJTw8DEzqEAAAAM/forsen-forsen-e.gif
</details>

### CODE REVIEW

**Comment by KararTY** [comment]
I've been unable to get avif to work, on Windows by using the suffix -all binary distribution, or on WSL/Ubuntu by rebuilding libvips with libheif-dev.

Should maybe still support 7TV previews? Their non-suffix/non-file-format-ending version uses animated WebP which works with out of the box libvips on Ubuntu & Windows -static/-all distributions.

**Comment by leon-richardt** [comment]
Should we maybe continue supporting the `enable-lilliput` parameter as an alias for `enable-animated-thumbnails` and print a deprecation warning when it's used? That way, this needn't be a breaking change.

**Comment by leon-richardt** [comment]
> > Should we maybe continue supporting the `enable-lilliput` parameter as an alias for `enable-animated-thumbnails` and print a deprecation warning when it's used? That way, this needn't be a breaking change.
> 
> I was thinking about making that change a separate PR, but since the switch to libvips is a breaking change and that feature hasn't landed in a released version of chatterino/api I thought it wouldn't really matter how many breaking changes you shove in. :joy: :ok_hand:

Good point. I agree now that the `enable-lilliput` should not be an alias.

> 
> I like the idea of a deprecation notice, however.
> 

I would even argue that we should error out if the `enable-lilliput` parameter is used in the config since ignoring it could lead to unexpected behavior. For example, suppose the following scenario:
1. User has `enable-lilliput: false` in his config to disable the generation of animated thumbnails.
2. User upgrades their version to the new release which uses `enable-animated-thumbnails` instead.
3. Since `enable-animated-thumbnails` is `true` by default, animated thumbnails will now be generated even though the user has not made any changes to their config.

By erroring out with a descriptive message when the old parameter is detected, we can make sure users are not running into unexpected behavior when upgrading.

File: pkg/thumbnail/thumbnail.go [hunk 1]
```diff
@@ -84,3 +84,42 @@ func BuildStaticThumbnail(inputBuf []byte, resp *http.Response) ([]byte, error)
...
+
+	// n=-1 is used for animated images to make sure to get all frames and not just the first one.
+	if format == vips.ImageTypeGIF || format == vips.ImageTypeWEBP {
+		importParams.NumPages.Set(-1)
+	}
+
+	image, err = vips.LoadThumbnailFromBuffer(inputBuf, maxThumbnailSize, maxThumbnailSize, vips.InterestingAll, vips.SizeDown, importParams)
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (leon-richardt) [inline pkg/thumbnail/thumbnail.go:112]: Should use a logger derived with logger.FromContext() here. (Requires passing down the request context to BuildAnimatedThumbnail() but this is considered good practice anyway.)

File: pkg/thumbnail/thumbnail.go [hunk 2]
```diff
@@ -84,3 +84,42 @@ func BuildStaticThumbnail(inputBuf []byte, resp *http.Response) ([]byte, error)
 
 	return outputBuf, nil
 }
+
+func BuildAnimatedThumbnail(inputBuf []byte, resp *http.Response) ([]byte, error) {
+	image, err := vips.NewImageFromBuffer(inputBuf)
+
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not load image from url: %s", resp.Request.URL)
+	}
```
Comment (leon-richardt) [inline pkg/thumbnail/thumbnail.go:93]: Should use a logger derived with logger.FromContext() here. (Requires passing down the request context to BuildAnimatedThumbnail() but this is considered good practice anyway.)

File: pkg/thumbnail/thumbnail.go [hunk 3]
```diff
@@ -84,3 +84,42 @@ func BuildStaticThumbnail(inputBuf []byte, resp *http.Response) ([]byte, error)
...
+		return []byte{}, fmt.Errorf("could not transform image from url: %s", resp.Request.URL)
+	}
+
+	// We export to WebP by default to save on bandwidth and cache storage.
+	exportParams := vips.NewWebpExportParams()
+	outputBuf, _, err := image.ExportWebp(exportParams)
+
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not export image from url: %s", resp.Request.URL)
+	}
```
Comment (leon-richardt) [inline pkg/thumbnail/thumbnail.go:122]: Should use a logger derived with logger.FromContext() here. (Requires passing down the request context to BuildAnimatedThumbnail() but this is considered good practice anyway.)

**Review by leon-richardt — COMMENTED** [review]
Other than changing the logging method to what's used elsewhere, I think this looks good :+1:

**Comment by KararTY** [comment]
> Other than changing the logging method to what's used elsewhere, I think this looks good 👍

Wouldn't this be a change for BuildStaticThumbnail() too? Maybe best if done in separate PR?

**Comment by leon-richardt** [comment]
> > Other than changing the logging method to what's used elsewhere, I think this looks good +1
> 
> Wouldn't this be a change for BuildStaticThumbnail() too? If so, maybe best if done in separate PR?

Yes, I should have done that in `BuildStaticThumbnail()` as well. But I still think that **this** PR should include the change for `BuildAnimatedthumbnail()`

File: pkg/thumbnail/thumbnail.go [hunk 4]
```diff
@@ -84,3 +86,46 @@ func BuildStaticThumbnail(inputBuf []byte, resp *http.Response) ([]byte, error)
...
+	// n=-1 is used for animated images to make sure to get all frames and not just the first one.
+	if format == vips.ImageTypeGIF || format == vips.ImageTypeWEBP {
+		importParams.NumPages.Set(-1)
+	}
+
+	image, err = vips.LoadThumbnailFromBuffer(inputBuf, maxThumbnailSize, maxThumbnailSize, vips.InterestingAll, vips.SizeDown, importParams)
+
+	if err != nil {
+		log.Errorw("could not transform image from url", "url", resp.Request.URL, "err", err)
+		return []byte{}, err
```
Comment (pajlada) [inline pkg/thumbnail/thumbnail.go:118]: user friendly error

File: pkg/thumbnail/thumbnail.go [hunk 5]
```diff
@@ -84,3 +86,46 @@ func BuildStaticThumbnail(inputBuf []byte, resp *http.Response) ([]byte, error)
...
+	}
+
+	// We export to WebP by default to save on bandwidth and cache storage.
+	exportParams := vips.NewWebpExportParams()
+	outputBuf, _, err := image.ExportWebp(exportParams)
+
+	if err != nil {
+		log.Errorw("could not export image from url", "url", resp.Request.URL, "err", err)
+		return []byte{}, err
```
Comment (pajlada) [inline pkg/thumbnail/thumbnail.go:127]: user friendly error

File: pkg/thumbnail/thumbnail.go [hunk 6]
```diff
@@ -84,3 +86,46 @@ func BuildStaticThumbnail(inputBuf []byte, resp *http.Response) ([]byte, error)
...
+	}
+
+	image, err = vips.LoadThumbnailFromBuffer(inputBuf, maxThumbnailSize, maxThumbnailSize, vips.InterestingAll, vips.SizeDown, importParams)
+
+	if err != nil {
+		log.Errorw("could not transform image from url", "url", resp.Request.URL, "err", err)
+		return []byte{}, err
+	}
+
+	// We export to WebP by default to save on bandwidth and cache storage.
```
Comment (pajlada) [inline pkg/thumbnail/thumbnail.go:121]: No need for a comment here, no justifications needed

**Review by pajlada — CHANGES_REQUESTED** [review]

**Review by pajlada — APPROVED** [review]

**Pull request merged**